File: scripts/ImageSubscriber.py
#!/usr/bin/python
import os
import rospy
from openhab_msgs.msg import ImageState
from sensor_msgs.msg import Image
from cv_bridge import CvBridge, CvBridgeError

# OpenCV
import cv2

VERBOSE = False

class ImageSubscriber(object):

    def __init__(self, item_name):
        self.item_name = item_name

        """Configure subscriber."""
        # Create a subscriber with appropriate topic, custom message and name of
        # callback function.
        self.sub = rospy.Subscriber("/openhab/items/" + self.item_name + "/state", ImageState, self.callback)

        # Initialize message variables.
        self.enable = False
        self.data = None
        self.bridge = CvBridge()
        self.image = None

        if self.enable:
            self.start()
        else:
            self.stop()

    def start(self):
        self.enable = True
        self.sub = rospy.Subscriber("/openhab/items/" + self.item_name + "/state", ImageState, self.callback)

    def callback(self, data):
        """Handle subscriber data."""

        self.data = data

        if self.data.isnull == False:
            msg = "Received %s" % self.data.item

            try:
                self.image = self.bridge.imgmsg_to_cv2(self.data.state, 'bgr8')
            except CvBridgeError as e:
                print(e)

            cv2.imshow(str(self.data.item), self.image)
            cv2.waitKey(25)
        else:
            msg = "Received %s with NULL" % self.data.item
        rospy.loginfo(rospy.get_caller_id() + msg)

    def stop(self):
        """Turn off subscriber."""
        self.enable = False
        self.sub.unregister()

if __name__ == '__main__':
    # Initialize the node and name it.
    node_name = "ImageSubscriberNode"
    rospy.init_node(node_name, anonymous=True)

    imageSubscriber = ImageSubscriber("testImage")

    # Go to the main loop
    try:
        imageSubscriber.start()
        # Wait for messages on topic, go to callback function when new messages arrive.
        # spin() simply keeps python from exiting until this node is stopped
        rospy.spin()
    # Stop with Ctrl + C
    except KeyboardInterrupt:
        imageSubscriber.stop()

        nodes = os.popen("rosnode list").readlines()
        for i in range(len(nodes)):
            nodes[i] = nodes[i].replace("\n","")

        for node in nodes:
            os.system("rosnode kill " + node_name)


        print("Node stopped")

File: src/ImageSubscriber.cpp
#include <ros/ros.h>
#include <opencv2/opencv.hpp>
#include <cv_bridge/cv_bridge.h>
#include <sensor_msgs/Image.h>
#include <openhab_msgs/ImageState.h>

bool VERBOSE = false;

class ImageSubscriber
{
public:
    ImageSubscriber(const std::string& item_name) : item_name_(item_name), enable_(false)
    {
        // Create a subscriber with appropriate topic, custom message and name of
        // callback function.
        sub_ = nh_.subscribe("/openhab/items/" + item_name_ + "/state", 1, &ImageSubscriber::callback, this);
    }

    void start()
    {
        enable_ = true;
        sub_ = nh_.subscribe("/openhab/items/" + item_name_ + "/state", 1, &ImageSubscriber::callback, this);
    }

    void callback(const openhab_msgs::ImageState::ConstPtr& data)
    {
        if (data->isnull == false)
        {
            ROS_INFO_STREAM("Received " << data->item);
            try
            {
                image_ = cv_bridge::toCvCopy(data->state, "bgr8")->image;
            }
            catch (const cv_bridge::Exception& e)
            {
                ROS_ERROR_STREAM("cv_bridge exception: " << e.what());
                return;
            }

            cv::imshow(item_name_, image_);
            cv::waitKey(25);
        }
        else
        {
            ROS_INFO_STREAM("Received " << data->item << " with NULL");
        }
    }

    void stop()
    {
        enable_ = false;
        sub_.shutdown();
    }

private:
    std::string item_name_;
    bool enable_;
    ros::NodeHandle nh_;
    ros::Subscriber sub_;
    cv::Mat image_;
};

int main(int argc, char** argv)
{
    // Initialize the node and name it.
    std::string node_name = "ImageSubscriberNode";
    ros::init(argc, argv, node_name, ros::init_options::AnonymousName);
    ros::NodeHandle nh;

    ImageSubscriber imageSubscriber("testImage");

    // Go to the main loop
    try
    {
        imageSubscriber.start();
        // Wait for messages on topic, go to callback function when new messages arrive.
        // spin() simply keeps ros from exiting until this node is stopped
        ros::spin();
    }
    // Stop with Ctrl + C
    catch (const std::exception&)
    {
        imageSubscriber.stop();

        std::vector<std::string> nodes;
        ros::master::getNodes(nodes);
        for (const auto& node : nodes)
        {
            ros::NodeHandle nh(node);
            std::string node_ns = nh.getNamespace();
            if (node_ns == ros::this_node::getNamespace())
            {
                ros::service::waitForService("rosout/get_loggers");
                ros::ServiceClient client = nh.serviceClient<rosgraph_msgs::Log>("rosout/get_loggers");
                rosgraph_msgs::Log log;
                log.request.level = rosgraph_msgs::Log::DEBUG;
                client.call(log);
                if (log.response.loggers[0].name == "/rosout")
                {
                    system(("rosnode kill " + node).c_str());
                }
            }
        }
        ROS_INFO("Node stopped");
    }

    return 0;
}
